Python dependency added to TorchVisionConfig.cmake despite building with -DUSE_PYTHON=off

### 🐛 Describe the bug

Despite building `libtorchvision.so` with the flag `-DUSE_PYTHON=off`, when building a program that depends on it (`infer-masts`), `cmake` throws an error due to a missing Python dependency:
```
$ Torch_DIR=~/.local/lib/python3.9/site-packages/torch/share/cmake/Torch cmake -DTORCH_CUDA_ARCH_LIST="5.3;6.2;7.2" -Wno-dev -GNinja ..
...
CMake Error at CMakeLists.txt:24 (add_executable):
  Target "infer-masts" links to target "Python3::Python" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
```
Changing line 61 of TorchVisionConfig.cmake from
```
set_target_properties(TorchVision::TorchVision PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${${PN}_INCLUDE_DIR}" INTERFACE_LINK_LIBRARIES "torch;Python3::Python" )
```
to
```
set_target_properties(TorchVision::TorchVision PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${${PN}_INCLUDE_DIR}" INTERFACE_LINK_LIBRARIES "torch" )
```
allows my program to build without issues.

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0a0+gitbc2c6ed
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (aarch64)
GCC version: (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.20.4
Libc version: glibc-2.27

Python version: 3.9.10 (main, Mar 16 2022, 15:59:07)  [GCC 8.4.0] (64-bit runtime)
Python platform: Linux-4.9.201-tegra-aarch64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/aarch64-linux-gnu/libcudnn.so.8.0.0
/usr/lib/aarch64-linux-gnu/libcudnn_adv_infer.so.8.0.0
/usr/lib/aarch64-linux-gnu/libcudnn_adv_train.so.8.0.0
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_infer.so.8.0.0
/usr/lib/aarch64-linux-gnu/libcudnn_cnn_train.so.8.0.0
/usr/lib/aarch64-linux-gnu/libcudnn_ops_infer.so.8.0.0
/usr/lib/aarch64-linux-gnu/libcudnn_ops_train.so.8.0.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] numpy-quaternion==2022.2.10.14.20.39
[pip3] torch==1.11.0a0+gitbc2c6ed
[pip3] torchvision==0.12.0a0+9b5a3fe
[conda] Could not collect
```
Note: This is on a Jetson Nano device. Python3.9, torch and torchvision were built with gcc-8, not gcc-7 as indicated by the collect_env.py script. Relevant environment variables are:
```
export BUILD_CAFFE2_OPS=OFF
export USE_FBGEMM=OFF
export USE_FAKELOWP=OFF
export BUILD_TEST=OFF
export USE_MKLDNN=OFF
export USE_NNPACK=OFF
export USE_XNNPACK=OFF
export USE_QNNPACK=OFF
export USE_PYTORCH_QNNPACK=OFF
export USE_CUDA=ON
export USE_CUDNN=ON
export TORCH_CUDA_ARCH_LIST="5.3;6.2;7.2"
export USE_NCCL=OFF
export USE_SYSTEM_NCCL=OFF
export USE_OPENCV=OFF
export MAX_JOBS=2
export PATH=/usr/lib/ccache:$PATH
export CC=gcc-8
export CXX=g++-8
```
torchvision was configured with the above environment variables and the command
```
torchvision-0.12.0/build$ cmake -Wno-dev -DWITH_CUDA=on -DTORCH_CUDA_ARCH_LIST="5.3;6.2;7.2" -DUSE_PYTHON=off -GNinja -DCMAKE_INSTALL_PREFIX=~/.local ..
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,6 +4,8 @@ set(CMAKE_CXX_STANDARD 14)
 file(STRINGS version.txt TORCHVISION_VERSION)
 
 option(WITH_CUDA "Enable CUDA support" OFF)
+option(WITH_PNG "Enable features requiring LibPNG." ON)
+option(WITH_JPEG "Enable features requiring LibJPEG." ON)
 option(USE_PYTHON "Link to Python when building" OFF)
 
 if(WITH_CUDA)
@@ -18,16 +20,22 @@ if(WITH_CUDA)
   endif()
 endif()
 
+find_package(Torch REQUIRED)
+
+if (WITH_PNG)
+    add_definitions(-DPNG_FOUND)
+    find_package(PNG REQUIRED)
+endif()
+
+if (WITH_JPEG)
+    add_definitions(-DJPEG_FOUND)
+    find_package(JPEG REQUIRED)
+endif()
+
 if (USE_PYTHON)
   add_definitions(-DUSE_PYTHON)
   find_package(Python3 REQUIRED COMPONENTS Development)
 endif()
-
-find_package(Torch REQUIRED)
-find_package(PNG REQUIRED)
-find_package(JPEG REQUIRED)
-add_definitions(-DJPEG_FOUND)
-add_definitions(-DPNG_FOUND)
 
 function(CUDA_CONVERT_FLAGS EXISTING_TARGET)
     get_property(old_flags TARGET ${EXISTING_TARGET} PROPERTY INTERFACE_COMPILE_OPTIONS)
@@ -82,7 +90,15 @@ FOREACH(DIR ${ALLOW_LISTED})
 ENDFOREACH()
 
 add_library(${PROJECT_NAME} SHARED ${ALL_SOURCES})
-target_link_libraries(${PROJECT_NAME} PRIVATE ${TORCH_LIBRARIES} ${PNG_LIBRARY} ${JPEG_LIBRARIES})
+target_link_libraries(${PROJECT_NAME} PRIVATE ${TORCH_LIBRARIES})
+
+if (WITH_PNG)
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${PNG_LIBRARY})
+endif()
+
+if (WITH_JPEG)
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${JPEG_LIBRARIES})
+endif()
 
 if (USE_PYTHON)
   target_link_libraries(${PROJECT_NAME} PRIVATE Python3::Python)
@@ -92,7 +108,15 @@ set_target_properties(${PROJECT_NAME} PROPERTIES
   EXPORT_NAME TorchVision
   INSTALL_RPATH ${TORCH_INSTALL_PREFIX}/lib)
 
-include_directories(torchvision/csrc ${JPEG_INCLUDE_DIRS} ${PNG_INCLUDE_DIRS})
+include_directories(torchvision/csrc)
+
+if (WITH_PNG)
+    include_directories(${PNG_INCLUDE_DIRS})
+endif()
+
+if (WITH_JPEG)
+    include_directories(${JPEG_INCLUDE_DIRS})
+endif()
 
 set(TORCHVISION_CMAKECONFIG_INSTALL_DIR "share/cmake/TorchVision" CACHE STRING "install path for TorchVisionConfig.cmake")
 

--- a/cmake/TorchVisionConfig.cmake.in
+++ b/cmake/TorchVisionConfig.cmake.in
@@ -22,23 +22,26 @@ if(NOT (CMAKE_VERSION VERSION_LESS 3.0))
 # Don't include targets if this file is being picked up by another
 # project which has already built this as a subproject
 #-----------------------------------------------------------------------------
-if(NOT TARGET ${PN}::TorchVision)
+if(NOT TARGET ${PN}::${PN})
 include("${CMAKE_CURRENT_LIST_DIR}/${PN}Targets.cmake")
 
-if(NOT TARGET torch_library)
+target_include_directories(${PN}::${PN} INTERFACE "${${PN}_INCLUDE_DIR}" ${JPEG_INCLUDE_DIRS} ${PNG_INCLUDE_DIRS})
+
 find_package(Torch REQUIRED)
-endif()
+find_package(PNG REQUIRED)
+find_package(JPEG REQUIRED)
+
+target_link_libraries(${PN}::${PN} INTERFACE torch ${PNG_LIBRARY} ${JPEG_LIBRARIES})
+
 if (@USE_PYTHON@)
   if(NOT TARGET Python3::Python)
     find_package(Python3 COMPONENTS Development)
   endif()
+  target_link_libraries(torch INTERFACE Python3::Python)
 endif()
 
-set_target_properties(TorchVision::TorchVision PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${${PN}_INCLUDE_DIR}" INTERFACE_LINK_LIBRARIES "torch;Python3::Python" )
-
-
 if(@WITH_CUDA@)
-  target_compile_definitions(TorchVision::TorchVision INTERFACE WITH_CUDA)
+  target_compile_definitions(${PN}::${PN} INTERFACE WITH_CUDA)
 endif()
 
 endif()

--- a/cmake/TorchVisionConfig.cmake.in
+++ b/cmake/TorchVisionConfig.cmake.in
@@ -25,23 +25,33 @@ if(NOT (CMAKE_VERSION VERSION_LESS 3.0))
 if(NOT TARGET ${PN}::${PN})
 include("${CMAKE_CURRENT_LIST_DIR}/${PN}Targets.cmake")
 
-target_include_directories(${PN}::${PN} INTERFACE "${${PN}_INCLUDE_DIR}" ${JPEG_INCLUDE_DIRS} ${PNG_INCLUDE_DIRS})
+target_include_directories(${PN}::${PN} INTERFACE "${${PN}_INCLUDE_DIR}")
+
+if(@WITH_CUDA@)
+  target_compile_definitions(${PN}::${PN} INTERFACE WITH_CUDA)
+endif()
 
 find_package(Torch REQUIRED)
-find_package(PNG REQUIRED)
-find_package(JPEG REQUIRED)
+target_link_libraries(${PN}::${PN} INTERFACE torch)
 
-target_link_libraries(${PN}::${PN} INTERFACE torch ${PNG_LIBRARY} ${JPEG_LIBRARIES})
+if(@WITH_PNG@)
+  find_package(PNG REQUIRED)
+  target_link_libraries(${PN}::${PN} INTERFACE ${PNG_LIBRARY})
+  target_compile_definitions(${PN}::${PN} INTERFACE PNG_FOUND)
+endif()
+
+if(@WITH_JPEG@)
+  find_package(JPEG REQUIRED)
+  target_link_libraries(${PN}::${PN} INTERFACE ${JPEG_LIBRARIES})
+  target_compile_definitions(${PN}::${PN} INTERFACE JPEG_FOUND)
+endif()
 
 if (@USE_PYTHON@)
   if(NOT TARGET Python3::Python)
     find_package(Python3 COMPONENTS Development)
   endif()
   target_link_libraries(torch INTERFACE Python3::Python)
-endif()
-
-if(@WITH_CUDA@)
-  target_compile_definitions(${PN}::${PN} INTERFACE WITH_CUDA)
+  target_compile_definitions(${PN}::${PN} INTERFACE USE_PYTHON)
 endif()
 
 endif()

